lua/entities/gmod_wire_expression2/base/optimizer.lua

```diff
@@ -0,0 +1,182 @@
+        if instruction[3][1] ~= 'literal' or instruction[4][1] ~= 'literal' then return instruction end
+        return evaluateBinary(instruction)
+    end
+end
+
+function constantPropagation.neg(instruction)
+    if instruction[3][1] ~= 'literal' then return instruction end
+    return evaluateUnary(instruction)
+end
+
+constantPropagation["not"] = function(instruction)
+    if instruction[3][1] ~= 'literal' then return instruction end
+    instruction[3] = evaluateUnary({'is', instruction[2], instruction[3]})
+    return evaluateUnary(instruction)
+end
+
+for _, operator in pairs({ "and", "or" }) do
+    constantPropagation[operator] = function(instruction)
+        if instruction[3][1] ~= 'literal' then return instruction end
+        instruction[3] = evaluateUnary({'is', instruction[2], instruction[3]})
+        instruction[4] = evaluateUnary({'is', instruction[2], instruction[4]})
+        return evaluateBinary(instruction)
+    end
+end
+
+table.insert(Optimizer.Passes, constantPropagation)
+
+
+local peephole = {}
+function peephole.add(instruction)
+    -- (add 0 x) → x
+    if instruction[3][1] == 'literal' and instruction[3][3] == 0 then return instruction[4] end
+    -- (add x 0) → x
+    if instruction[4][1] == 'literal' and instruction[4][3] == 0 then return instruction[3] end
+    -- (add (neg x) (neg y)) → (neg (add x y))
+    if instruction[3][1] == 'neg' and instruction[4][1] == 'neg' then
+        return {'neg', instruction[2], {'add', instruction[2], instruction[3][3], instruction[4][3],
+                __instruction = true}}
+    end
+    -- (add x (neg y)) → (sub x y)
+    if instruction[4][1] == 'neg' then
+        return {'sub', instruction[2], instruction[3], instruction[4][3]}
+    end
+    -- (add (neg x) y) → (sub y x)
+    if instruction[3][1] == 'neg' then
+        return {'sub', instruction[2], instruction[4], instruction[3][3]}
+    end
+    return instruction
+end
+
+function peephole.sub(instruction)
+    -- (sub 0 x) → (neg x)
+    if instruction[3][1] == 'literal' and instruction[3][3] == 0 then
+        return {'neg', instruction[2], instruction[4]}
+    end
+    -- (sub x 0) → x
+    if instruction[4][1] == 'literal' and instruction[4][3] == 0 then return instruction[3] end
+    -- (sub (neg x) (neg y)) → (sub y x)
+    if instruction[3][1] == 'neg' and instruction[4][1] == 'neg' then
+        return {'sub', instruction[2], instruction[4][3], instruction[3][3]}
+    end
+    -- (sub x (neg y) → (add x y))
+    if instruction[4][1] == 'neg' then
+        return {'add', instruction[2], instruction[3], instruction[4][3]}
+    end
+    -- (sub (neg x) y) → (neg (add x y))
+    if instruction[3][1] == 'neg' then
+        return {'neg', instruction[2], {'add', instruction[2], instruction[3][3], instruction[4],
+                __instruction = true }}
+    end
+    return instruction
+end
+
+function peephole.mul(instruction)
+    if instruction[4][1] == 'literal' and instruction[3][1] ~= 'literal' then
```

Comment:
This same trick should be used in `peephole.add`

Reply:
That actually doesn't make the code any shorter - it's beneficial for `mul` because there are three identities (`(mul 1 x) → x`, `(mul 0 x) → 0`, `(mul -1 x) → (neg x)`), whereas `add` only has one.

Commutivity of `add` might come in handy in future, when reordering eg. `x + 1 + y + 2 → x + y + 3`, but that's not implemented at the moment.